File: telo/src/test/java/com/hartwig/hmftools/telo/analysers/TelomericSplitReadAnalyserTest.java
package com.hartwig.hmftools.telo.analysers;

import static com.hartwig.hmftools.telo.TeloUtils.hasTelomericContent;

import com.hartwig.hmftools.telo.TeloUtils;
import com.hartwig.hmftools.telo.breakend.TelomericBreakEnd;
import com.hartwig.hmftools.telo.breakend.TelomericSplitReadAnalyser;

import static junit.framework.TestCase.assertFalse;
import static junit.framework.TestCase.assertTrue;
import static junit.framework.TestCase.assertEquals;

import org.junit.Test;

import htsjdk.samtools.SAMRecord;

public class TelomericSplitReadAnalyserTest
{
    @Test
    public void testFindTelomericBreakEnd()
    {
        TelomericSplitReadAnalyser analyser = new TelomericSplitReadAnalyser();
        analyser.setBoundaryZone(2);
        analyser.setMinSoftClipLength(10);

        SAMRecord record = new SAMRecord(null);
        record.setReadName("TestRead");
        record.setReferenceName("1");
        record.setAlignmentStart(1000);
        record.setCigarString("20S30M");
        record.setMateReferenceName("2");
        record.setMateAlignmentStart(200);
        record.setReadPairedFlag(true);
        record.setFirstOfPairFlag(true);

        // positive strand
        record.setReadNegativeStrandFlag(false);

        // first 18 is telomeric
        record.setReadString("TAACCCTAACCCTAACCC" + "AGCTAGCTAGCTACGTTGCAACCTGACGAA");

        TelomericBreakEnd tbe = analyser.findTelomericBreakEnd(record);

        assertEquals(TelomericBreakEnd.Type.LEFT_C_TELOMERIC, tbe.getType());
        assertEquals("1", tbe.getChromosome());
        assertEquals(999, tbe.getPosition());

        // try the negative strand
        record.setReadNegativeStrandFlag(true);
        record.setReadString(TeloUtils.reverseComplementSequence(record.getReadString()));

        tbe = analyser.findTelomericBreakEnd(record);

        assertEquals(TelomericBreakEnd.Type.LEFT_C_TELOMERIC, tbe.getType());
        assertEquals("1", tbe.getChromosome());
        assertEquals(999, tbe.getPosition());

        // right soft clip
        record.setCigarString("30M20S");
        // last 18 is telomeric
        record.setReadString("AGCTAGCTAGCTACGTTGCAACCTGACGAA" + "TAACCCTAACCCTAACCC");

        record.setReadNegativeStrandFlag(false);
        tbe = analyser.findTelomericBreakEnd(record);

        assertEquals(TelomericBreakEnd.Type.RIGHT_C_TELOMERIC, tbe.getType());
        assertEquals("1", tbe.getChromosome());
        assertEquals(1030, tbe.getPosition());

        record.setReadNegativeStrandFlag(true);
        record.setReadString(TeloUtils.reverseComplementSequence(record.getReadString()));
        tbe = analyser.findTelomericBreakEnd(record);

        assertEquals(TelomericBreakEnd.Type.RIGHT_C_TELOMERIC, tbe.getType());
        assertEquals("1", tbe.getChromosome());
        assertEquals(1030, tbe.getPosition());

        // test G telomeric
        record.setReadNegativeStrandFlag(false);
        record.setReadString("TTAGGGTTAGGGTTAGGG" + "AGCTAGCTAGCTACGTTGCAACCTGACGAA");
        record.setCigarString("20S30M");
        tbe = analyser.findTelomericBreakEnd(record);

        assertEquals(TelomericBreakEnd.Type.LEFT_G_TELOMERIC, tbe.getType());
        assertEquals("1", tbe.getChromosome());
        assertEquals(999, tbe.getPosition());
    }

    @Test
    public void testLikelyTelomeric()
    {
        String readBasesG = "AGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGTGTTAGGG";
        String readBasesC = "CTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAACCTAACCCAAACCCTAACCCAAACCCTAACCCTAACCCTAAC";

        assertTrue(TeloUtils.isLikelyGTelomeric(readBasesG));
        assertFalse(TeloUtils.isLikelyCTelomeric(readBasesG));

        assertFalse(TeloUtils.isLikelyGTelomeric(readBasesC));
        assertTrue(TeloUtils.isLikelyCTelomeric(readBasesC));

        //readBasesC = "CCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAACCAAACCCAAACCCTAACCCAAACCCACACCCCCACACCAACCCCCACCCCCACCACAACACCCCCCCCCCCCCCCCCCCCACC";

    }



}

File: teal/src/test/java/com/hartwig/hmftools/teal/analysers/TelomericSplitReadAnalyserTest.kt
package com.hartwig.hmftools.teal.analysers

import com.hartwig.hmftools.teal.breakend.SplitReadAnalyser
import htsjdk.samtools.SAMRecord
import com.hartwig.hmftools.teal.breakend.*
import com.hartwig.hmftools.teal.TeloUtils
import junit.framework.TestCase
import org.junit.Test

class TelomericSplitReadAnalyserTest
{
    @Test
    fun testFindTelomericBreakEnd()
    {
        val analyser = SplitReadAnalyser(0.9, 0.8, 10)
        analyser.setBoundaryZone(2)
        //analyser.setMinSoftClipLength(10)
        val record = SAMRecord(null)
        record.readName = "TestRead"
        record.referenceName = "1"
        record.alignmentStart = 1000
        record.cigarString = "20S30M"
        record.mateReferenceName = "2"
        record.mateAlignmentStart = 200
        record.readPairedFlag = true
        record.firstOfPairFlag = true

        // positive strand
        record.readNegativeStrandFlag = false

        // first 18 is telomeric
        record.readString = "TAACCCTAACCCTAACCC" + "AGCTAGCTAGCTACGTTGCAACCTGACGAA"
        var tbe = analyser.findTelomericBreakEnd(record)
        TestCase.assertEquals(TelomericBreakEndType.LEFT_C_TELOMERIC, tbe!!.type)
        TestCase.assertEquals("1", tbe.chromosome)
        TestCase.assertEquals(999, tbe.position)

        // try the negative strand
        record.readNegativeStrandFlag = true
        record.readString = TeloUtils.reverseComplementSequence(record.readString)
        tbe = analyser.findTelomericBreakEnd(record)
        TestCase.assertEquals(TelomericBreakEndType.LEFT_C_TELOMERIC, tbe!!.type)
        TestCase.assertEquals("1", tbe.chromosome)
        TestCase.assertEquals(999, tbe.position)

        // right soft clip
        record.cigarString = "30M20S"
        // last 18 is telomeric
        record.readString = "AGCTAGCTAGCTACGTTGCAACCTGACGAA" + "TAACCCTAACCCTAACCC"
        record.readNegativeStrandFlag = false
        tbe = analyser.findTelomericBreakEnd(record)
        TestCase.assertEquals(TelomericBreakEndType.RIGHT_C_TELOMERIC, tbe!!.type)
        TestCase.assertEquals("1", tbe.chromosome)
        TestCase.assertEquals(1030, tbe.position)
        record.readNegativeStrandFlag = true
        record.readString = TeloUtils.reverseComplementSequence(record.readString)
        tbe = analyser.findTelomericBreakEnd(record)
        TestCase.assertEquals(TelomericBreakEndType.RIGHT_C_TELOMERIC, tbe!!.type)
        TestCase.assertEquals("1", tbe.chromosome)
        TestCase.assertEquals(1030, tbe.position)

        // test G telomeric
        record.readNegativeStrandFlag = false
        record.readString = "TTAGGGTTAGGGTTAGGG" + "AGCTAGCTAGCTACGTTGCAACCTGACGAA"
        record.cigarString = "20S30M"
        tbe = analyser.findTelomericBreakEnd(record)
        TestCase.assertEquals(TelomericBreakEndType.LEFT_G_TELOMERIC, tbe!!.type)
        TestCase.assertEquals("1", tbe.chromosome)
        TestCase.assertEquals(999, tbe.position)
    }

    @Test
    fun testLikelyTelomeric()
    {
        val readBasesG = "AGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGGGTTAGTGTTAGGG"
        val readBasesC = "CTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAACCTAACCCAAACCCTAACCCAAACCCTAACCCTAACCCTAAC"
        TestCase.assertTrue(TeloUtils.isLikelyGTelomeric(readBasesG))
        TestCase.assertFalse(TeloUtils.isLikelyCTelomeric(readBasesG))
        TestCase.assertFalse(TeloUtils.isLikelyGTelomeric(readBasesC))
        TestCase.assertTrue(TeloUtils.isLikelyCTelomeric(readBasesC))

        //readBasesC = "CCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAACCAAACCCAAACCCTAACCCAAACCCACACCCCCACACCAACCCCCACCCCCACCACAACACCCCCCCCCCCCCCCCCCCCACC";
    }
}